Stripe/STPTelemetryClient.m

```diff
@@ -0,0 +1,188 @@
+//
+//  STPTelemetryClient.m
+//  Stripe
+//
+//  Created by Ben Guo on 4/18/17.
+//  Copyright © 2016 Stripe, Inc. All rights reserved.
+//
+
+@import WebKit;
+@import UIKit;
+#import <sys/utsname.h>
+
+#import "STPTelemetryClient.h"
+#import "STPAPIClient.h"
+
+typedef NS_ENUM(NSInteger, STPMetricField) {
+    STPMetricFieldCookieSupport = 0,
+    STPMetricFieldDoNotTrack,
+    STPMetricFieldLanguage,
+    STPMetricFieldPlatform,
+    STPMetricFieldPlugins,
+    STPMetricFieldScreenSize,
+    STPMetricFieldTimeZoneOffset,
+    STPMetricFieldTouchSupport,
+    STPMetricFieldAvailableStorage,
+    STPMetricFieldFonts,
+    STPMetricFieldGraphicsConfiguration,
+    STPMetricFieldUserAgent,
+    STPMetricFieldFlashVersion,
+    STPMetricFieldHasAdBlocker,
+    STPMetricFieldCanvasId,
+    STPMetricFieldMax
+};
+
+@interface STPTelemetryClient ()
+@property (nonatomic) NSDate *appOpenTime;
+@property (nonatomic, readwrite) NSURLSession *urlSession;
+@property (nonatomic, readwrite) NSString *userAgent;
+@property (nonatomic, readwrite) WKWebView *webView;
+@end
+
+@implementation STPTelemetryClient
+
++ (BOOL)shouldSendTelemetry {
+#if TARGET_OS_SIMULATOR
+    return NO;
+#endif
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunreachable-code"
+    return NSClassFromString(@"XCTest") == nil;
+#pragma clang diagnostic pop
+}
+
++ (instancetype)sharedInstance {
+    static id sharedClient;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedClient = [self new];
+    });
+    return sharedClient;
+}
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
+        _urlSession = [NSURLSession sessionWithConfiguration:config];
+        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(applicationDidBecomeActive) name:UIApplicationDidBecomeActiveNotification object:nil];
+        WKWebView *webView = [WKWebView new];
+        [webView loadHTMLString:@"<html></html>" baseURL:nil];
+        [webView evaluateJavaScript:@"navigator.userAgent" completionHandler:^(id result, __unused NSError *error) {
+            self.userAgent = result;
+        }];
+        _webView = webView;
+    }
+    return self;
+}
+
+- (void)dealloc {
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)addTelemetryFieldsToParams:(NSMutableDictionary *)params {
+    params[@"muid"] = [self muid];
+    params[@"time_on_page"] = [self timeOnPage];
+}
+
+- (void)applicationDidBecomeActive {
+    self.appOpenTime = [NSDate date];
+}
+
+- (NSString *)muid {
+    NSString *muid = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
+    return muid ?: @"";
+}
+
+- (NSNumber *)timeOnPage {
```

Comment:
Yeah – we want to have a time_on_page value for users who don't use our UI components too. On the web, this is "number of milliseconds between Stripe.js loading and the user submitting card information" – I think app open time should provide a pretty similar signal.